comms/dht/src/network_discovery/seed_strap.rs

```diff
@@ -156,326 +158,97 @@ impl SeedStrap {
+            round_info.sync_peers = candidates.iter().map(|p| p.node_id.clone()).collect();
+            if candidates.is_empty() {
+                debug!(target: LOG_TARGET, "SeedStrap: No more seed peer candidates available to try.");
+                break;
+            } else {
+                debug!(
                     target: LOG_TARGET,
-                    "SeedStrap: Iteration {}/{}: Seed peer '{}' returned an empty peer list. Disconnecting.",
-                    idx + 1,
-                    num_seeds_to_try,
-                    seed_peer_node_id_str
+                    "SeedStrap: Preparing to sync from up to {} seed peers. Selected peer IDs for this round: {:?}",
+                    num_seeds_this_round,
+                    round_info.sync_peers,
                 );
-                if let Err(e) = conn.disconnect(Minimized::Yes).await {
-                    warn!(target: LOG_TARGET, "SeedStrap: Failed to disconnect from seed peer '{}' after receiving empty list: {}", seed_peer_node_id_str, e);
-                }
-                continue;
             }
+            remaining_seed_peers = remaining_seed_peers.saturating_sub(candidates.len());
+
+            // Get peers from seeds concurrently
+            let mut join_handles = Vec::new();
+            for (idx, seed_peer_candidate) in candidates.into_iter().enumerate() {
+                attempted_seed_contacts += 1;
+
+                let seed_node_ids_set_clone = seed_node_ids_set.clone();
+                let context_clone = self.context.clone();
+
+                let handle = tokio::task::spawn(async move {
+                    get_peers(
+                        context_clone,
+                        seed_peer_candidate,
+                        num_seeds_this_round,
+                        idx,
+                        &seed_node_ids_set_clone,
+                        max_peers_to_sync_per_round,
+                        max_permitted_peer_claims,
+                        max_permitted_peer_addresses_per_claim,
+                    )
+                    .await
+                });
 
-            // This seed successfully provided peers
-            round_info.num_succeeded += 1;
-            successful_seed_contacts += 1;
-
-            debug!(
-                target: LOG_TARGET,
-                "SeedStrap: Successfully fetched {} peer entries from seed peer '{}'. Disconnecting before processing.",
-                peers_from_seed.len(),
-                seed_peer_node_id_str
-            );
-
-            if let Err(e) = conn.disconnect(Minimized::Yes).await {
-                warn!(target: LOG_TARGET, "SeedStrap: Failed to disconnect from seed peer '{}': {}", seed_peer_node_id_str, e);
-            } else {
-                debug!(target: LOG_TARGET, "SeedStrap: Successfully disconnected from seed peer '{}'", seed_peer_node_id_str);
+                join_handles.push(handle);
             }
+            let results = future::join_all(join_handles).await;
```

Comment:
this needs to be changed

Reply:
> We can process the results as they com back, we don't need to await all to be completed before we begin processing, use JoinSet

This is not what is happening. `async fn get_peers(` is spawned concurrently for each seed peer that needs to be contacted. The spawned function will do all that is required - it will fetch/stream the peer info list from the seed peer and process the data into the peer db. We `join_all(join_handles).await` all spawned processes to complete, see how many failed, then spawn new concurrent processes for the ones that failed, until the quota is met or the seed peer list is depleted.

The only processing that needs to happen after the `join_all(join_handles).await` is the processing of stats.

Reply:
Code updated to use `FuturesUnordered` and `stream.next().await`